installer/cmd/render-mirror.go

```diff
@@ -0,0 +1,208 @@
+// See License-AGPL.txt in the project root for license information.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/docker/distribution/reference"
+	"github.com/gitpod-io/gitpod/installer/pkg/common"
+	"github.com/spf13/cobra"
+	"regexp"
+	"sigs.k8s.io/yaml"
+	"sort"
+	"strings"
+)
+
+type mirrorRepo struct {
+	Original string `json:"original"`
+	Target   string `json:"target"`
+}
+
+var mirrorOpts struct {
+	ExcludeThirdParty bool
+	Output            string
+}
+
+var outputTypes = []string{"yaml", "json"}
+
+// mirrorCmd represents the mirror command
+var mirrorCmd = &cobra.Command{
+	Use:   "mirror",
+	Short: "Renders a list of images used so they can be mirrored to a third-party registry",
+	Long: `Renders a list of images used so they can be mirrored to a third-party registry
+
+A config file is required which can be generated with the init command.
+
+The output can then be used to iterate over each image. A script can
+be written to pull from the "original" path and then tag and push the
+image to the "target" repo`,
+	Example: `
+  # YAML
+  gitpod-installer render mirror --config config.yaml > mirror.yaml
+
+  # JSON
+  gitpod-installer render mirror --config config.yaml -o json > mirror.json
+
+  # Pull original and push to target
+  for row in $(gitpod-installer render mirror --config ./config.yaml ./versions.yaml -o json | jq -c '.[]'); do
+    original=$(echo $row | jq -r '.original')
+    target=$(echo $row | jq -r '.target')
+    docker pull $original
+    docker tag $original $target
+    docker push $target
+  done`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// Validate output type
+		var output string
+		for _, i := range outputTypes {
+			if mirrorOpts.Output == i {
+				output = i
+			}
+		}
+		if output == "" {
+			return fmt.Errorf("unknown output type: %s", mirrorOpts.Output)
+		}
+
+		_, cfgVersion, cfg, err := loadConfig(renderOpts.ConfigFN)
+		if err != nil {
+			return err
+		}
+
+		// Get the target repository from the config
+		targetRepo := strings.TrimRight(cfg.Repository, "/")
+
+		// Use the default Gitpod registry to pull from
+		cfg.Repository = common.GitpodContainerRegistry
+
+		k8s, err := renderKubernetesObjects(cfgVersion, cfg)
+		if err != nil {
+			return err
+		}
+
+		// Map of images used for deduping
+		allImages := make(map[string]bool)
+
+		rawImages := make([]string, 0)
+		for _, item := range k8s {
+			rawImages = append(rawImages, getPodImages(item)...)
+			rawImages = append(rawImages, getGenericImages(item)...)
+		}
+
+		images := make([]mirrorRepo, 0)
+		for _, img := range rawImages {
+			// Dedupe
+			if _, ok := allImages[img]; ok {
+				continue
+			}
+			allImages[img] = true
+
+			// Convert target
+			target := img
+			if strings.Contains(img, cfg.Repository) {
+				// This is the Gitpod registry
+				target = strings.Replace(target, cfg.Repository, targetRepo, 1)
+			} else if !mirrorOpts.ExcludeThirdParty {
```

Comment:
When `--exclude-third-party` is set, the output still contains the third-party image but the target is not replaced (`original` and `target` is the same in the output). This is not inherently bad, but `docker push` (as in the sample bash for loop) would fail for these images. It also surprised me at first. What do you think: Should we remove these third-party images entirely from the output?

Reply:
Yeah, that makes much more sense